src/hist.jl

```diff
@@ -237,3 +239,150 @@ fit{N}(::Type{Histogram},vs::NTuple{N,AbstractVector}, wv::WeightVec; closed::Sy
     closed = _check_closed_arg(closed,:fit)
     fit(Histogram, vs, wv, histrange(vs,nbins,closed); closed=closed)
 end
+
+
+# Get a suitable high-precision type for the norm of a histogram.
+@generated function norm_type{T, N, E}(h::Histogram{T, N, E})
+    args = [:( eltype(edges[$d]) ) for d = 1:N]
+    quote
+        edges = h.edges
+        norm_type(promote_type(T, $(args...)))
+    end
+end
+
+norm_type{T<:Integer}(::Type{T}) = promote_type(T, Int64)
+norm_type{T<:AbstractFloat}(::Type{T}) = promote_type(T, Float64)
+
+
+"""
+    norm(h::Histogram)
+
+Calculate the norm of histogram `h` as the absolute value of its integral.
+"""
+@generated function norm{T, N, E}(h::Histogram{T, N, E})
+    quote
+        edges = h.edges
+        weights = h.weights
+        S = norm_type(h)
+        v_0 = 1
+        s_0 = zero(S)
+        @inbounds @nloops(
+            $N, i, weights,
+            d -> begin
+                v_{$N-d+1} = v_{$N-d} * (edges[d][i_d + 1] - edges[d][i_d])
+                s_{$N-d+1} = zero(S)
+            end,
+            d -> begin
+                s_{$N-d} += s_{$N-d+1}
+            end,
+            begin
+                $(Symbol("s_$(N)")) += (@nref $N weights i) * $(Symbol("v_$N"))
+            end
+        )
+        norm(s_0)
+    end
+end
+
+
+function _float_deepcopy(x)
+    y = float(x)
+    y === x? deepcopy(y) : y
+end
+
+
+function float{T, N, E}(h::Histogram{T, N, E})
```

Comment:
> We don't want to make defensive copies.  Copy only when it is needed or we end up like R. 

Ok, then I suggest we keep the currently implemented behavior. But if we copy, it should be a *full* copy (as implemented now), not a partial one. Otherwise it'll be completely unpredictable for the user - and then users will do exactly what you don't want: Make lot's of defensive copies in their code to get thread-safety.